src/fomo.jl

```diff
@@ -0,0 +1,399 @@
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+solver = FomoSolver(nlp);
+stats = solve!(solver, nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+## `R2`
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+stats = R2(nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+solver = FoSolver(nlp);
+stats = solve!(solver, nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+"""
+mutable struct FomoSolver{T, V} <: AbstractFirstOrderSolver
+  x::V
+  g::V
+  c::V
+  m::V
+  d::V
+  p::V
+  α::T
+end
+
+function FomoSolver(nlp::AbstractNLPModel{T, V}) where {T, V}
+  x = similar(nlp.meta.x0)
+  g = similar(nlp.meta.x0)
+  c = similar(nlp.meta.x0)
+  m = fill!(similar(nlp.meta.x0), 0)
+  d = fill!(similar(nlp.meta.x0), 0)
+  p = similar(nlp.meta.x0)
+  return FomoSolver{T, V}(x, g, c, m, d, p, T(0))
+end
+
+@doc (@doc FomoSolver) function fomo(nlp::AbstractNLPModel{T, V}; kwargs...) where {T, V}
+  solver = FomoSolver(nlp)
+  solver_specific = Dict(:avgβmax => T(0.))
+  stats = GenericExecutionStats(nlp;solver_specific=solver_specific)
+  return solve!(solver, nlp, stats; kwargs...)
+end
+
+function SolverCore.reset!(solver::FomoSolver{T}) where {T}
+  fill!(solver.m,0)
+  solver
+end
+
+SolverCore.reset!(solver::FomoSolver, ::AbstractNLPModel) = reset!(solver)
+
+mutable struct FoSolver{T, V} <: AbstractFirstOrderSolver
+  x::V
+  g::V
+  c::V
+  α::T
+end
+
+function FoSolver(nlp::AbstractNLPModel{T, V}) where {T, V}
+  x = similar(nlp.meta.x0)
+  g = similar(nlp.meta.x0)
+  c = similar(nlp.meta.x0)
+  return FoSolver{T, V}(x, g, c, T(0))
+end
+
+@doc (@doc FomoSolver) function R2(nlp::AbstractNLPModel{T, V}; kwargs...) where {T, V}
+  solver = FoSolver(nlp)
+  stats = GenericExecutionStats(nlp)
+  if haskey(kwargs,:σmin)
+    return solve!(solver, nlp, stats; step_backend = r2_step(), αmax = 1/kwargs[:σmin], kwargs...)
+  else
+    return solve!(solver, nlp, stats; step_backend = r2_step(), kwargs...)
+  end
+end
+
+function SolverCore.reset!(solver::FoSolver{T}) where {T}
+  solver
+end
+
+SolverCore.reset!(solver::FoSolver, ::AbstractNLPModel) = reset!(solver)
+
+function SolverCore.solve!(
+  solver::AbstractFirstOrderSolver,
+  nlp::AbstractNLPModel{T, V},
+  stats::GenericExecutionStats{T, V};
+  callback = (args...) -> nothing,
+  x::V = nlp.meta.x0,
+  atol::T = √eps(T),
+  rtol::T = √eps(T),
+  η1::T = T(eps(T)^(1 / 4)),
+  η2::T = T(0.95),
+  γ1::T = T(1/2),
+  γ2::T = T(2),
+  γ3::T = T(1/2),
+  αmax::T = 1/eps(T),
+  max_time::Float64 = 30.0,
+  max_eval::Int = -1,
+  max_iter::Int = typemax(Int),
+  β::T = T(0.9),
+  θ1::T = T(0.1),
+  θ2::T = T(eps(T)^(1/3)),
+  verbose::Int = 0,
+  step_backend = r2_step(),
+  σmin = nothing # keep consistency with R2 interface. kwargs immutable, can't delete it in `R2`
+) where {T, V}
+  use_momentum = typeof(solver) <: FomoSolver
```

Comment:
What happens when we write another `FirstOrderSolver <: AbstractFirstOrderSolver` that doesn’t use the algorithm below??? There will be a clash. The argument should be `AbstractFomoSolver <: AbstractFirstOrderSolver`, I think.

Reply:
Exact. Would you suggest to define the new abstract type `AbstractFomoSolver` or simply set the arguement to `solver <: Union{FoSolver,FomoSolver}`

Reply:
If the union works, it’s a solution.